core/domain/exp_domain.py

```diff
@@ -474,6 +474,285 @@ def __init__(self, change_list):
         }
 
 
+class StateVersionSpan(python_utils.OBJECT):
+    """A consecutive span of versions in which a state remains "equivalent".
+
+    Equivalence between spans are defined using a user-defined predicate, passed
+    in to the constructor. By default, all states are equivalent to each other.
+
+    Spans must be extended in increasing version-order. It is an error to extend
+    a span with a version older than what it already contains; exceptions are
+    raised to enforce this contract.
+    """
+
+    def __init__(
+            self, exp_version, state_name, state,
+            state_equality_predicate=None):
+        """Initializes a span for the given version of an exploration's state.
+
+        Args:
+            exp_version: int. The version the state corresponds to.
+            state_name: str. The name of the state at the given version.
+            state: state_domain.State. The structure of the state at the given
+                version.
+            state_equality_predicate:
+                callable(state_domain.State, state_domain.State) -> bool | None.
+                Returns True when two states are "equal". The predicate is used
+                to enforce invariant: all states in a span are equivalent.
+                If None, then states will always be equivalent to each other.
+        """
+        self._version_start, self._version_end = exp_version, exp_version + 1
+        self._version_snapshots = (
+            collections.OrderedDict({exp_version: (state_name, state)}))
+        self._state_equality_predicate = (
+            state_equality_predicate or
+            StateVersionSpan._default_state_equality_predicate)
+
+    def __iter__(self):
+        """Returns an iterable over the snapshots held by the span."""
+        return iter(self._version_snapshots.items())
+
+    def __contains__(self, version):
+        """Returns whether the given version is covered by self."""
+        return self._version_start <= version < self._version_end
+
+    def __getitem__(self, version):
```

Comment:
Nevermind, the reduced code duplication is worthwhile IMO so I brought it back. To compensate, I've added comments explaining what the syntactic sugar resolves to.

Reply:
While what you currently have is probably the best implementation of a dict mapping, this is the first time we've done this sort of thing in this codebase (and we have other models that represent mappings, see e.g. TrainingJobExplorationMappingModel). As is usual with infrastructural conventions, I think we should introduce that for all models at once, or not at all, rather than have two different patterns in the codebase.

Is there actually an issue with using self._get_version() or similar?

Reply:
I think the thing that confuses me is I don't totally understand yet how this & the ExplorationStatesHistory domain objects are meant to be used. Is there a design document or one pager that explains this in more detail?

Per the PR description, I understand how these objects could be beneficial for recomputing stats, but I don't yet have a good handle on what the end result will look like. Will these only ever be used in MR/cron jobs? If so, it may be better not to complicate the domain layer of explorations by introducing these objects (since that also invites them to be used in other ways).

I'm specifically digging into how these be used because I wonder whether it would be simpler to compute the runs of equivalent versions at the moment we need it. I think that might result in an overall simpler implementation, potentially not even requiring new objects at all and mitigate Sean's original concern here. How does this compare with what you're thinking @brianrodri?

Reply:
So I plan to use this in MR jobs that map over ExplorationModel instances. Such maps always return the latest versions of the exploration, not any of the prior versions, so these domain objects bundle all that fetching into a single `exp_services` operation.

Does this address your concern?

Reply:
@brianrodri not exactly, but it provides more context. All new domain objects are public API--we're basically inviting other services to leverage them for future functionality. I'm not sure that we actually want that here since this is a complex operation with very intended use. By including it as part of the public domain API, we will need to maintain it over time, and this bit of machinery may not be straightforward to maintain.

If we plan to reuse this in multiple jobs, then maybe a service-only method that returns a simple dict may be more straightforward to maintain. Even if that gets used elsewhere other than the jobs, it will be less complex to update if version mapping changes in the future (note that versioning semantics will change as a requirement for the Android app, and it's not yet clear how).

If this is only ever needed for a single job, I suggest using a standard dict directly where it's needed rather than changing the domain layer public API at all.

Can you clarify what benefits we get with this object vs. a standard dict?

Reply:
Got it.

So the primary benefit this class provides is answering the question: "What's the lifespan of this State?" I discussed a bit with @seanlip offline, and we believe this does have uses beyond just stats, it should be _the_ programmatic API for understanding how/when a state has changed over time.

If the versioning semantics are planned to change, I think we should make the API resistant to them and perhaps have it answer "more vague" questions. WDYT?

Reply:
Thanks @brianrodri & @seanlip for the additional examples and discussions. I agree that this does make sense as a permanent addition to the domain API. (For some brief context on my original hesitation, there will be major changes to version management across all activities next year as part of the Android project & I was concerned that this may result in us switching patterns multiple times, but I don't think that's actually a major concern anymore after discussing this further).

That being said, I also agree with @seanlip that I prefer not to have this object treated like a dict itself. That may be a nice pattern to move toward, but I think we should consider it as a codebase-wide pattern rather than introduce it as part of this change. I think there are very specific pros/cons when viewing the pattern in that context which may change perspective on whether this is a pattern that we want to pursue). Rather than blocking on those considerations, it probably would be easier to move forward with an API that matches other existing APIs (e.g. with simple getters/setters as needed).

Reply:
OK, I've updated the API to use a `get` and `get_multi` family of methods instead. PTAL!